### PR TITLE
Id token verification fixes

### PR DESCRIPTION
**Pull request by m0n9oose**
Incude:
- #34 
- #36 

Please take a look @januszm, @krzysiek1507 & @Eric-Guo

### CODE REVIEW

File: lib/omniauth/strategies/openid_connect.rb [hunk 1]
```diff
@@ -313,14 +310,29 @@ def id_token_callback_phase
...
 
-        error_attrs = RESPONSE_TYPE_EXCEPTIONS[options.response_type.to_s]
+        error_attrs = RESPONSE_TYPE_EXCEPTIONS[response_type]
         fail!(error_attrs[:key], error_attrs[:exception_class].new(params['error']))
 
         false
       end
 
+      def response_type
+        @response_type ||= options.response_type.to_s
```
Comment (krzysiek1507) [inline lib/omniauth/strategies/openid_connect.rb]:
```ruby
[2] pry(main)> ['id_token', 'code'].to_s
=> "[\"id_token\", \"code\"]"
```

I don't think that it's a good response_type. It should be like: `id_token,code`

File: lib/omniauth/strategies/openid_connect.rb [hunk 2]
```diff
@@ -313,14 +310,29 @@ def id_token_callback_phase
...
+      end
+
+      def id_token_response?
+        Array(response_type).include?('id_token')
```
Comment (krzysiek1507) [inline lib/omniauth/strategies/openid_connect.rb]: Array from a string?

File: lib/omniauth/strategies/openid_connect.rb [hunk 3]
```diff
@@ -313,14 +310,29 @@ def id_token_callback_phase
       end
 
       def valid_response_type?
-        return true if params.key?(options.response_type)
+        return true if params.key?(response_type)
```
Comment (krzysiek1507) [inline lib/omniauth/strategies/openid_connect.rb]: `id_token,code` as key?

File: lib/omniauth/strategies/openid_connect.rb (hunk 1 above)
Comment (januszm) [reply lib/omniauth/strategies/openid_connect.rb]: valid response type params are separated by single space character, e.g. `id_token token`

**Comment by krzysiek1507** [comment]
I wanted to add multiple response types once #34 is merged

**Comment by januszm** [comment]
@m0n9oose  I'd rather close this one and merge / review both mine and @krzysiek1507 PRs separately. One baby step at a time.

File: lib/omniauth/strategies/openid_connect.rb (hunk 1 above)
Comment (krzysiek1507) [reply lib/omniauth/strategies/openid_connect.rb]: Oh, sorry, that's true! Too many standards!

Comment (januszm) [reply lib/omniauth/strategies/openid_connect.rb]: We need to make a clear distinction between this gem's setting `options.response_type` which, from what I see, is not yet determined for multiple response types and the OAuth2 standard for response_type (which is clearly defined: one or multiple values, separated with spaces).

File: lib/omniauth/strategies/openid_connect.rb (hunk 2 above)
Comment (m0n9oose) [reply lib/omniauth/strategies/openid_connect.rb]: This is just rewrite of your own code https://github.com/m0n9oose/omniauth_openid_connect/pull/34/files#diff-8467ba16957a2513a82eb9fe781131dfR186

Comment (krzysiek1507) [reply lib/omniauth/strategies/openid_connect.rb]: No, it's not. I use `options.response_type` and you use `response_type` method which returns a string.

**Comment by m0n9oose** [comment]
> Close the combined PR please.

Sorry I didn’t get what’s the difference? This PR is an exact copy of #34 and #36 with some minor changes (extracted ‘response_type’ method to avoid ‘to_s’ as you mentioned in your PR and cleaner ‘id_token_response?’) in two separate commits. It’s your code with my cosmetic changes.

Now you want me to close this PR, but you don’t have time to work on your PR.

File: lib/omniauth/strategies/openid_connect.rb (hunk 2 above)
Comment (m0n9oose) [reply lib/omniauth/strategies/openid_connect.rb]: And? Yes, we cast response type to string to make comparisons consistent. How does it correspond to your first question about an array?

Comment (krzysiek1507) [reply lib/omniauth/strategies/openid_connect.rb]: How it is going to work?
```ruby
[6] pry(main)> Array(['id_token', 'code'].to_s).include?('id_token')
=> false
```

Comment (m0n9oose) [reply lib/omniauth/strategies/openid_connect.rb]: > How it is going to work?
> 
> ```ruby
> [6] pry(main)> Array(['id_token', 'code'].to_s).include?('id_token')
> => false
> ```

Ah, okay, I see what you mean. Good catch, thanks

**Comment by Eric-Guo** [comment]
Glad to see two developers working in this issue, I will testing after you ping me.

**Comment by januszm** [comment]
@m0n9oose  I said that if you have suggestions how to change my PR, write in comments and I can apply them there if they bring value. If you don't like the solution or have your own idea how to fix it, close my PR and start new. It's not the best practice to ship different features or changes in one PR, like that.

**Pull request closed without merging**